proposals/roles.md

```diff
@@ -0,0 +1,169 @@
+﻿# Roles
+
+This proposal is divided into three parts, all relating to extending existing types:
+A. static extensions
+B. roles and extensions with members
+C. roles and extensions that implement interfaces
+
+The syntax for a roles is as follows:
+
+```antlr
+type_declaration
+    : role_declaration // add
+    | extension_declaration // add
+    | ...
+    ;
+
+role_declaration
+    : modifier* 'role' identifier type_parameter_list? role_underlying_type type_parameter_constraints_clause* role_body
+    ;
+
+extension_declaration
+    : modifier* 'extension' identifier type_parameter_list? role_underlying_type type_parameter_constraints_clause* role_body
+    ;
+
+role_underlying_type
+    : ':' type
+    : ':' type ',' interface_type_list
+    ;
+
+role_body
+    : '{' role_member_declaration* '}'
+    ;
+
+role_member_declaration
+    : constant_declaration
+    | field_declaration
+    | method_declaration
+    | property_declaration
+    | event_declaration
+    | indexer_declaration
+    | operator_declaration
+    | type_declaration
+    ;
+```
+
+## Phase A: Adding static constants, fields, methods and properties
+
+In this first subset of the feature, the syntax is restricted to non-nested **extension_declaration**
+and containing only **constant_declaration**, **field_declaration**, **method_declaration** and **property_declaration** members.
+TODO: events?
+
+### Extension type
+
+The permitted modifiers on an extension type are `partial`, `unsafe` and the accessibility modifiers `public` and `internal`.  
```

Comment:
Does it have to match or be no more broad than the accessibility of the underlying role type?

Reply:
Added. Thanks